elfi/testbench/testbench.py

```diff
@@ -0,0 +1,295 @@
+        repetitions : int
+            How many repetitions of models is included in the testbench.
+        observation : np.array, optional
+            Observation, if available.
+        reference_parameter : dictionary, optional
+            True parameter values if available.
+        reference_posterior : np.array, optional
+            A sample from a reference posterior.
+        progress_bar : boolean
+            Indicate whether to display testbench progressbar.
+        seed : int, optional
+
+        """
+        self.model = model
+        self.method_list = []
+        self.method_seed_list = []
+        self.repetitions = repetitions
+        self.rng = np.random.RandomState(seed)
+        self.observations = observations
+        self.reference_parameter = reference_parameter
+        # TODO Add functionality to deal with reference posterior
+        self.reference_posterior = reference_posterior
+        self.simulator_name = list(model.observed)[0]
+        if progress_bar:
+            self.progress_bar = ProgressBar(prefix='Progress', suffix='Complete',
+                                            decimals=1, length=50, fill='=')
+        else:
+            self.progress_bar = None
+
+        self._resolve_test_type()
+        self._collect_tests()
+
+    def _collect_tests(self):
+        self.test_dictionary = {
+            'model': self.model,
+            'observations': self.observations,
+            'reference_parameter': self.reference_parameter,
+            'reference_posterior': self.reference_posterior
+        }
+
+    def _get_seeds(self, n_rep=1):
+        """Fix a seed for each of the repeated instances."""
+        upper_limit = 2 ** 32 - 1
+        return self.rng.randint(
+            low=0,
+            high=upper_limit,
+            size=n_rep,
+            dtype=np.uint32)
+
+    def _resolve_test_type(self):
+        self._set_default_test_type()
+        self._resolve_reference_parameters()
+        self._resolve_observations()
+
+    def _set_default_test_type(self):
+        self.description = {
+            'observations_available': self.observations is not None,
+            'reference_parameters_available': self.reference_parameter is not None,
+            'reference_posterior_available': self.reference_posterior is not None
+            }
+
+    def _resolve_reference_parameters(self):
+        if self.description['reference_parameters_available']:
+            for keys, values in self.reference_parameter.items():
+                self.reference_parameter[keys] = np.repeat(
+                    values,
+                    repeats=self.repetitions
+                    )
+
+        else:
+            seed = self._get_seeds(n_rep=1)
+            self.reference_parameter = self.model.generate(
+                batch_size=self.repetitions,
+                outputs=self.model.parameter_names,
+                seed=seed[0])
+
+        self.param_dim = len(self.reference_parameter)
```

Comment:
i would propose not to sample reference parameters when `observations_available` is `True` and the parameters will not be used (unless user calls error calculation in which case these random parameter values are compared to the estimates calculated based on the observations). 

`self.param_dim = len(self.model.parameter_names)` could be set in init so reference parameters would not be needed for that and could be none.

Reply:
Modified by removing random `reference_parameter` generation when observations are available. 

Also now we set `self.param_dim` in the constructor.